Mathlib/RingTheory/AdicCompletion/Algebra.lean

```diff
@@ -59,11 +64,61 @@ def subalgebra : Subalgebra R (∀ n, R ⧸ (I ^ n • ⊤ : Ideal R)) :=
 def subring : Subring (∀ n, R ⧸ (I ^ n • ⊤ : Ideal R)) :=
   Subalgebra.toSubring (subalgebra I)
 
-instance : CommRing (AdicCompletion I R) :=
-  inferInstanceAs <| CommRing (subring I)
-
-instance : Algebra R (AdicCompletion I R) :=
-  inferInstanceAs <| Algebra R (subalgebra I)
+/-- One in `AdicCompletion I R`. -/
+@[irreducible]
+def one : AdicCompletion I R :=
+  ⟨1, by simp⟩
+
+instance : One (AdicCompletion I R) where
+  one := one I
+
+/-- Multiplication in `AdicCompletion I R`. -/
+@[irreducible]
+def mul (x y : AdicCompletion I R) : AdicCompletion I R :=
+  ⟨x.val * y.val, by simp [x.property, y.property]⟩
+
+instance : Mul (AdicCompletion I R) where
+  mul := mul I
+
+section
+
+unseal mul
```

Comment:
For future reference, this partially defeats the purpose of `irreducible`. You can do a more targeted `unseal X in` in a proof step I believe. Or you can wrap your term in `with_unfolding_all` to achieve a similar effect

Reply:
I thought that the `unseal` here is confined to this section? But you are saying that even this is too big of a scope?

Reply:
You are correct, it is. But with the definition open, Lean will try to do extra work there. Less than overall, but still can be avoided. 

The pattern is usually: 
- make `@[irreducible] def X` 
- develop minimal necessary API for `X` either `unseal`ing or `with_all_unfolding`
- never `unseal` again 

So in general, I would never see an `unseal X` unless it ended in an `in`

Reply:
I see. But in the case above where I need `unseal` to see that the two `AdicCompletion I R`-module instances on `AdicCompletion I R` are defeq, I seem to be bound to use `unseal` whenever they come up. Is this correct or can I make this work somehow?

Reply:
If you doing it by hand, then yes you are stuck with that. If it is happening in instance synthesis, it is much worse.